server/src/main/java/org/elasticsearch/search/aggregations/metrics/weighted_avg/InternalWeightedAvg.java

```diff
@@ -0,0 +1,140 @@
+import org.elasticsearch.search.aggregations.pipeline.PipelineAggregator;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+public class InternalWeightedAvg extends InternalNumericMetricsAggregation.SingleValue implements WeightedAvg {
+    private final double sum;
+    private final double weight;
+
+    public InternalWeightedAvg(String name, double sum, double weight, DocValueFormat format, List<PipelineAggregator> pipelineAggregators,
+                               Map<String, Object> metaData) {
+        super(name, pipelineAggregators, metaData);
+        this.sum = sum;
+        this.weight = weight;
+        this.format = format;
+    }
+
+    /**
+     * Read from a stream.
+     */
+    public InternalWeightedAvg(StreamInput in) throws IOException {
+        super(in);
+        format = in.readNamedWriteable(DocValueFormat.class);
+        sum = in.readDouble();
+        weight = in.readDouble();
+    }
+
+    @Override
+    protected void doWriteTo(StreamOutput out) throws IOException {
+        out.writeNamedWriteable(format);
+        out.writeDouble(sum);
+        out.writeDouble(weight);
+    }
+
+    @Override
+    public double value() {
+        return getValue();
+    }
+
+    @Override
+    public double getValue() {
+        return sum / weight;
+    }
+
+    double getSum() {
+        return sum;
+    }
+
+    double getWeight() {
+        return weight;
+    }
+
+    DocValueFormat getFormatter() {
+        return format;
+    }
+
+    @Override
+    public String getWriteableName() {
+        return WeightedAvgAggregationBuilder.NAME;
+    }
+
+    @Override
+    public InternalWeightedAvg doReduce(List<InternalAggregation> aggregations, ReduceContext reduceContext) {
+        double weight = 0;
+        double sum = 0;
+        double sumCompensation = 0;
+        double weightCompensation = 0;
+        // Compute the sum of double values with Kahan summation algorithm which is more
+        // accurate than naive summation.
+        for (InternalAggregation aggregation : aggregations) {
+            InternalWeightedAvg avg = (InternalWeightedAvg) aggregation;
+            if (Double.isFinite(avg.weight) == false) {
+                weight += avg.weight;
+            } else if (Double.isFinite(weight)) {
+                double corrected = avg.weight - weightCompensation;
+                double newWeight = weight + corrected;
+                weightCompensation = (newWeight - weight) - corrected;
+                weight = newWeight;
+            }
+            if (Double.isFinite(avg.sum) == false) {
+                sum += avg.sum;
+            } else if (Double.isFinite(sum)) {
+                double corrected = avg.sum - sumCompensation;
+                double newSum = sum + corrected;
+                sumCompensation = (newSum - sum) - corrected;
+                sum = newSum;
+            }
```

Comment:
Could you add comments to the loop to explain why wee need each of the conditions?

Reply:
I'm actually not entirely sure.  @jpountz, are the conditionals to keep the naive behavior with infinites?  E.g. if an infinite is added it converts the final value to infinite, whereas kahan summing would do something different?

So it's basically bwc with how we did things before?